api/metrics/v1/labels_enforcer.go

```diff
@@ -0,0 +1,144 @@
+		// Adapted from
+		// https://github.com/prometheus-community/prom-label-proxy/blob/952266db4e0b8ab66b690501e532eaef33300596/injectproxy/routes.go.
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			id, ok := authentication.GetTenantID(r.Context())
+			if !ok {
+				http.Error(w, "error finding tenant ID", http.StatusInternalServerError)
+
+				return
+			}
+
+			e := injectproxy.NewEnforcer([]*labels.Matcher{{
+				Name:  label,
+				Type:  labels.MatchEqual,
+				Value: id,
+			}}...)
+			enforceRequestQueryLabels(e, w, r)
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+// WithEnforceAuthorizationLabels return a middleware that ensures every query
+// has a set of labels returned by the OPA authorizer enforced.
+func WithEnforceAuthorizationLabels() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			data, ok := authentication.GetAuthorizationData(r.Context())
+			if !ok {
+				http.Error(w, "error finding authorization label matcher", http.StatusInternalServerError)
+
+				return
+			}
+
+			// Early pass to the next if no authz
+			// label enforcement configured.
+			if data == "" {
+				next.ServeHTTP(w, r)
+
+				return
+			}
+
+			var lm []*labels.Matcher
+			if err := json.Unmarshal([]byte(data), &lm); err != nil {
+				http.Error(w, "error parsing authorization label matcher", http.StatusInternalServerError)
+
+				return
+			}
+
+			e := injectproxy.NewEnforcer(lm...)
+			enforceRequestQueryLabels(e, w, r)
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func enforceRequestQueryLabels(e *injectproxy.Enforcer, w http.ResponseWriter, r *http.Request) {
+	// The `query` can come in the URL query string and/or the POST body.
+	// For this reason, we need to try to enforcing in both places.
+	// Note: a POST request may include some values in the URL query string
+	// and others in the body. If both locations include a `query`, then
+	// enforce in both places.
+	q, found1, err := enforceQueryValues(e, r.URL.Query())
+	if err != nil {
+		http.Error(w, fmt.Sprintf("could not enforce labels: %v", err), http.StatusInternalServerError)
+
+		return
+	}
+
+	r.URL.RawQuery = q
+
+	var found2 bool
+	// Enforce the query in the POST body if needed.
+	if r.Method == http.MethodPost {
+		if err := r.ParseForm(); err != nil {
+			http.Error(w, fmt.Sprintf("could not parse form: %v", err), http.StatusInternalServerError)
+
+			return
+		}
+
+		q, found2, err = enforceQueryValues(e, r.PostForm)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("could not enforce labels: %v", err), http.StatusInternalServerError)
+
+			return
+		}
+		// We are replacing request body, close previous one (ParseForm ensures it is read fully and not nil).
+		_ = r.Body.Close()
+		r.Body = ioutil.NopCloser(strings.NewReader(q))
+		r.ContentLength = int64(len(q))
+	}
+
+	// If no query was found, return early.
+	if !found1 && !found2 {
```

Comment:
Just for me to understand better, why is it an error if no query is found for metrics labels enforcer but not for logs labels enforcer?

Reply:
AFAIU, the metrics handler requires the presence of a query to enforce it's labels. I bet this is a line of defense for empty queries.